include/cuco/priority_queue.cuh

```diff
@@ -0,0 +1,238 @@
+*
+* The device-side operations allow a cooperative group to push or pop from
+* device code. These device side
+* operations are invoked with a trivially-copyable device view,
+* `device_mutable_view` which can be obtained with the host function
+* `get_mutable_device_view` and passed to the device.
+*
+* @tparam T Type of the elements stored in the queue
+* @tparam Compare Comparison operator used to order the elements in the queue
+* @tparam FavorInsertionPerformance When true, insertion performance is
+*         increased at the expense of deletion performance.
+* @tparam Allocator Allocator defining how memory is allocated internally
+*/
+template <typename T, typename Compare = thrust::less<T>,
+          bool FavorInsertionPerformance = false,
+          typename Allocator = cuco::cuda_allocator<char>>
+class priority_queue {
+
+  using int_allocator_type = typename std::allocator_traits<Allocator>
+                                ::rebind_alloc<int>;
+
+  using t_allocator_type = typename std::allocator_traits<Allocator>
+                                          ::rebind_alloc<T>;
+
+  using size_t_allocator_type = typename std::allocator_traits<Allocator>
+                                          ::rebind_alloc<size_t>;
+
+  const int NodeSize = FavorInsertionPerformance ? 64 : 1024;
+
+ public:
+  /**
+   * @brief Construct a priority queue
+   *
+   * @param initial_capacity The number of elements the priority queue can hold
+   * @param alloc Allocator used for allocating device storage
+   */
+  priority_queue(size_t initial_capacity,
+                  Allocator const& alloc = Allocator{});
+
+  /**
+   * @brief Push elements into the priority queue
+   *
+   * @tparam InputIt Device accessible input iterator whose `value_type`
+   *        can be converted to T
+   * @param first Beginning of the sequence of elements
+   * @param last End of the sequence of elements
+   * @param stream The stream in which the underlying device operations will be
+   *               executed
+   */
+  template <typename InputIt>
+  void push(InputIt first, InputIt last, cudaStream_t stream = 0);
+
+  /**
+   * @brief Remove a sequence of the lowest elements ordered by Compare
+   *
+   * @tparam OutputIt Device accessible output iterator whose `value_type`
+   *        can be converted to T
+   * @param first Beginning of the sequence of output elements
+   * @param last End of the sequence of output elements
+   * @param stream The stream in which the underlying GPU operations will be
+   *               run
+   */
+  template <typename OutputIt>
+  void pop(OutputIt first, OutputIt last, cudaStream_t stream = 0);
+
+  /*
+  * @brief Return the amount of shared memory required for operations
+  *        on the queue with a thread block size of block_size
+  *
+  * @param block_size Size of the blocks to calculate storage for
+  * @return The amount of temporary storage required in bytes
+  */
+  int get_shmem_size(int block_size) {
```

Comment:
When is this needed by a user? i.e., does it need to be a public function?

Reply:
This is needed by a user who wishes to use the device view, as the amount of shared memory required by operations depends on the size of the blocks used for the operation. This could be avoided by always requiring the amount of shared memory for the maximum block size, if that is better.

Reply:
Another option would be to have a shared memory object like in CUB. In that case, we could either require that the block size be known at compile time or assume the maximum block size as above.